src/tree_list_model.rs

```diff
@@ -138,62 +150,76 @@ where Item: Copy,
-                    .upgrade().ok_or(ModelError::ParentDropped)?;
-                let mut node_parent = node_parent_ref.borrow_mut();
+        let node_parent_ref = node.parent
+            .as_ref().ok_or(ModelError::ParentNotSet)?
+            .upgrade().ok_or(ModelError::ParentDropped)?;
+        let mut node_parent = node_parent_ref.borrow_mut();
 
-                // Add this node to the parent's list of expanded child nodes.
-                if expanded {
-                    node_parent.children.insert(node.item_index, node_ref.clone());
-                } else {
-                    node_parent.children.remove(&node.item_index);
-                }
-            }
+        // Add this node to the parent's list of expanded child nodes.
+        if expanded {
+            node_parent.children.insert(node.item_index, node_ref.clone());
+        } else {
+            node_parent.children.remove(&node.item_index);
+        }
 
-            // Traverse back up the tree, modifying `child_count` for expanded/collapsed entries.
-            let mut position = node.relative_position()?;
-            let mut current_node = node_ref.clone();
-            while let Some(parent_weak) = current_node.clone().borrow().parent.as_ref() {
-                let parent = parent_weak.upgrade().ok_or(ModelError::ParentDropped)?;
-                if expanded {
-                    parent.borrow_mut().child_count += node.child_count;
-                } else {
-                    parent.borrow_mut().child_count -= node.child_count;
-                }
-                current_node = parent;
-                position += current_node.borrow().relative_position()? + 1;
-            }
+        drop(node_parent);
 
+        // Traverse back up the tree, modifying `total_child_count` for expanded/collapsed entries.
+        let mut position = node.relative_position()?;
+        let mut current_node = node_ref.clone();
+        while let Some(parent_weak) = current_node.clone().borrow().parent.as_ref() {
+            let parent = parent_weak.upgrade().ok_or(ModelError::ParentDropped)?;
             if expanded {
-                model.items_changed(position, 0, node.child_count);
+                parent.borrow_mut().total_child_count += node.total_child_count;
             } else {
-                model.items_changed(position, node.child_count, 0);
+                parent.borrow_mut().total_child_count -= node.total_child_count;
             }
+            current_node = parent;
+            position += current_node.borrow().relative_position()? + 1;
+        }
+
+        if expanded {
+            model.items_changed(position, 0, node.total_child_count);
+        } else {
+            model.items_changed(position, node.total_child_count, 0);
         }
 
-        node_ref.borrow_mut().expanded = expanded;
         Ok(())
     }
 
+    pub fn update(&mut self) -> Option<(u32, u32, u32)> {
+        let mut cap = self.capture.lock().unwrap();
```

Comment:
I realise this gets called from an the Glib idle handler which is an awkward place to handle errors, but let's try to keep things as clean as we can - we can still return a `Result` here, and put the `unwrap` in the idle handler.